File: src/main/java/io/github/takenoha/towerdefense/paper/ResourceVoucherItemData.java
package io.github.takenoha.towerdefense.paper;

import io.github.takenoha.towerdefense.persistence.ResourceType;
import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

/** Parsed, untrusted PDC view of a physical resource voucher. */
public record ResourceVoucherItemData(
        UUID voucherId,
        UUID teamId,
        ResourceType resourceType,
        long quantity,
        Optional<UUID> deliveryOperationId,
        Optional<UUID> redeemOperationId) {
    public ResourceVoucherItemData {
        Objects.requireNonNull(voucherId, "voucherId");
        Objects.requireNonNull(teamId, "teamId");
        ResourceType.require(resourceType);
        Objects.requireNonNull(deliveryOperationId, "deliveryOperationId");
        Objects.requireNonNull(redeemOperationId, "redeemOperationId");
        if (quantity <= 0L) {
            throw new IllegalArgumentException("voucher quantity must be positive");
        }
        if (deliveryOperationId.isPresent() && redeemOperationId.isPresent()) {
            throw new IllegalArgumentException("a voucher cannot have two receipt states");
        }
    }

    public boolean hasReceipt() {
        return deliveryOperationId.isPresent() || redeemOperationId.isPresent();
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/paper/ResourceVoucherItemData.kt
package io.github.takenoha.towerdefense.paper

import io.github.takenoha.towerdefense.persistence.ResourceType
import java.util.Objects
import java.util.Optional
import java.util.UUID
import kotlin.jvm.JvmRecord

/** Parsed, untrusted PDC view of a physical resource voucher. */
@JvmRecord
data class ResourceVoucherItemData(
    val voucherId: UUID,
    val teamId: UUID,
    val resourceType: ResourceType,
    val quantity: Long,
    val deliveryOperationId: Optional<UUID>,
    val redeemOperationId: Optional<UUID>,
) {
    init {
        Objects.requireNonNull(voucherId, "voucherId")
        Objects.requireNonNull(teamId, "teamId")
        ResourceType.require(resourceType)
        Objects.requireNonNull(deliveryOperationId, "deliveryOperationId")
        Objects.requireNonNull(redeemOperationId, "redeemOperationId")
        require(quantity > 0L) { "voucher quantity must be positive" }
        require(!(deliveryOperationId.isPresent && redeemOperationId.isPresent)) {
            "a voucher cannot have two receipt states"
        }
    }

    fun hasReceipt(): Boolean = deliveryOperationId.isPresent || redeemOperationId.isPresent
}
